src/for.jl

```diff
@@ -0,0 +1,97 @@
+using Distributions
+import Distributions.logpdf
+using Base.Cartesian
+
+
+export logpdf
+export rand
+
+export For
+struct For{F,T,D,X} 
+    f :: F  
+    θ :: T
+end
+
+#########################################################
+# T <: NTuple{N,J} where {J <: Integer}
+#########################################################
+
+For(f, θ::J...) where {J <: Integer} = For(f,θ)
+
+function For(f::F, θ::T) where {F, N, J <: Integer, T <: NTuple{N,J}}
+    d = f.(ones(Int, N)...)
```

Comment:
What if different indices return different distributions?

Reply:
We can check this without running the code using `Core.Compiler.return_type`, e.g. `Core.Compiler.return_type(+, Tuple{Int, Int}) === Int`. If inference fails then we can assume the output is anything.

Reply:
The assumption here is that this isn't the case, in order to avoid dynamic dispatch. My suggestion would be for the user to handle multiple return distributions by instead always returning a mixture.

But I'd never seen `Core.Compiler.return_type`, that looks really useful!

Reply:
I think with `Core.Compiler.return_type`, one can support the single and multiple return distribution types simultaneously. The former will be inferred and use static dispatch while the latter provides an additional feature for a performance hit which some users might still want in the experimental phase of their project before optimizing the code's performance.

Reply:
This is great in principle, but it's pretty strange that there's no multi-argument method:

```julia
julia> methods(Core.Compiler.return_type)
# 2 methods for generic function "return_type":
[1] return_type(f, t) in Core.Compiler at compiler/typeinfer.jl:626
[2] return_type(f, t, world) in Core.Compiler at compiler/typeinfer.jl:631
```

Maybe I'm missing something?

Reply:
The second argument to `return_type` goes `Tuple{arg1T, arg2T, ...}`, see my `+` example above.